tests/Console/E2ETest.php

```diff
@@ -0,0 +1,230 @@
+<?php
+/**
+ * Copyright © 2017-2018 Maks Rafalko
+ *
+ * License: https://opensource.org/licenses/BSD-3-Clause New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Console;
+
+use Infection\Command\ConfigureCommand;
+use Infection\Console\Application;
+use Infection\Console\InfectionContainer;
+use Infection\Finder\ComposerExecutableFinder;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Console\Input\ArgvInput;
+use Symfony\Component\Console\Output\BufferedOutput;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Process\Process;
+
+class E2ETest extends TestCase
+{
+    const EXPECT_ERROR = 1;
+    const EXPECT_SUCCESS = 0;
+
+    private $cwd;
+
+    /**
+     * @var \Composer\Autoload\ClassLoader|null
+     */
+    private $previousLoader;
+
+    protected function setUp()
+    {
+        // E2E tests usually require to chdir to their location
+        // Hence we would need to go back to this dir
+        $this->cwd = \getcwd();
+    }
+
+    /**
+     * Longest test: runs under about 160-200 sec
+     *
+     * To be run with:
+     *
+     * php -dmemory_limit=128M vendor/bin/phpunit --group=large
+     *
+     * @group e2e
+     * @large
+     */
+    public function test_it_runs_on_itself()
+    {
+        $this->requiresMemoryLimitSet();
+
+        $output = $this->runInfection(self::EXPECT_SUCCESS, [
+            '--test-framework-options="--exclude-group=e2e"',
+        ]);
+
+        $this->assertRegExp('/\d+ mutations were generated/', $output);
+        $this->assertRegExp('/\d{2,} mutants were killed/', $output);
+    }
+
+    /**
+     * @group e2e
+     */
+    public function test_it_runs_configure_command_if_no_configuration()
+    {
+        chdir('tests/Fixtures/e2e/Unconfigured/');
+
+        $output = $this->runInfection(self::EXPECT_ERROR);
+
+        $this->assertContains(ConfigureCommand::NONINTERACTIVE_MODE_ERROR, $output);
+    }
+
+    /**
+     * @dataProvider e2eTestSuiteDataProvider
+     * @group e2e
+     */
+    public function test_it_runs_an_e2e_test_with_success(string $dirName)
+    {
+        $this->runOnE2EFixture($dirName);
+    }
+
+    public function e2eTestSuiteDataProvider(): \Generator
+    {
+        $directories = Finder::create()
+            ->depth('== 0')
+            ->in(__DIR__ . '/../Fixtures/e2e/')
+            ->directories();
+
+        foreach ($directories as $dirName) {
+            if (file_exists($dirName . '/run_tests.bash')) {
+                // skipping non-standard tests
+                // specifically Memory_Limit - it is very slow to fail
+                continue;
+            }
+
+            yield [
+                $dirName,
+            ];
+        }
+    }
+
+    protected function tearDown()
+    {
+        if ($this->previousLoader) {
+            $this->previousLoader->unregister();
+        }
+
+        \chdir($this->cwd);
```

Comment:
can we drop those leading backslashes? It's not a performance critical code and I think it just hurts readability in general. If we want to I recommend to import it via a `use` statement instead which can be configured in IDEs (Editor -> General -> Auto-Imports in PhpStorm). And without an automated step, like with a CS Fixer (and the one for the backslashes is only for a few opcache optimised functions and doesn't support the import), we're bound to miss some (like in the method just below)

Reply:
Good point. CS Fixer can add these leading backslashes all over the place if needed, but I don't think we need them here or anywhere, I'll remove them.